### PR TITLE
feat: implemented OIDC presentation submission API

### LINKED ISSUE #433: [OpenID4VP] Implement OIDC Submit Presentation API in wallet-sdk

### PR DESCRIPTION
**Pull request by birtony**
Fixes #433

Signed-off-by: Anton Biriukov <anton.biriukov@avast.com>

### CODE REVIEW

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js [hunk 1]
```diff
@@ -120,6 +129,182 @@ export class OpenID4VP {
...
+        "Error submitting OpenID4VP presentation: expiry is missing"
+      );
+    }
+
+    const header = new Object();
+    Object.defineProperty(header, "alg", {
+      value: alg,
+    });
+    Object.defineProperty(header, "kid", {
+      value: issuer,
```
Comment (sudeshrshetty) [inline cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:185]: setting issuer in place of `kid`?

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js [hunk 2]
```diff
@@ -120,6 +129,182 @@ export class OpenID4VP {
...
+
+    const vp = new Object();
+    Object.defineProperty(payload, "@context", {
+      value: presentationQuery["@context"],
+    });
+    Object.defineProperty(payload, "type", {
+      value: presentationQuery.type,
+    });
+    Object.defineProperty(payload, "verifiableCredential", {
+      value: presentationQuery.verifiableCredential,
```
Comment (sudeshrshetty) [inline cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:205]: '**base64urlencoded**' verifiable credential

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js [hunk 3]
```diff
@@ -120,6 +129,182 @@ export class OpenID4VP {
...
+      );
+    } else if (!presentationQuery.verifiableCredential) {
+      throw new TypeError(
+        "Error submitting OpenID4VP presentation: verifiableCredential is missing"
+      );
+    } else if (!issuer) {
+      throw new TypeError(
+        "Error submitting OpenID4VP presentation: issuer is missing"
+      );
+    } else if (!expiry) {
```
Comment (sudeshrshetty) [inline cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:174]: Are you sure `exp` in mandatory?

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js [hunk 4]
```diff
@@ -120,6 +129,182 @@ export class OpenID4VP {
...
+      "Error generating ID Token: presentationSubmission cannot be empty"
+    );
+  } else if (!header) {
+    throw new TypeError("Error generating ID Token: header cannot be empty");
+  } else if (!expiry) {
+    throw new TypeError("Error generating ID Token: expiry cannot be empty");
+  }
+
+  const vpToken = new Object();
+  Object.defineProperty(vpToken, "presentation_submission", {
```
Comment (sudeshrshetty) [inline cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:244]: just curious about this syntax,

Why can't

``
var vpToken = {}
vpToken. presentation_submission = presentationSubmission

``

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 1 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:185]: If I grasped it correctly - https://identity.foundation/jwt-vc-presentation-profile/#exp-jwt-claim

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 2 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:205]: Right, maybe we could open an issue for it in aries? This shouldn't require any changes here once aries code is updated

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 3 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:174]: If I grasped it correctly - https://identity.foundation/jwt-vc-presentation-profile/#exp-jwt-claim

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 1 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js]: Fixed

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 4 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:245]: This way, the properties inside the `vpToken` object are not modifiable

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 3 above)
Comment (sudeshrshetty) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:174]: but it doesn't say it is mandatory

Comment (sudeshrshetty) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:174]: wordings r bit confusing, lets leave it as it for now.

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 2 above)
Comment (sudeshrshetty) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:206]: why we need to have issue in aries, you are directly creating presentation here, you are not using aries.

You just have to change line #206

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 4 above)
Comment (birtony) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:245]: I cannot use this syntax to create property `@context` on the object, which is required for https://github.com/trustbloc/agent-sdk/pull/453/files#diff-73c781567add49e4fe09e28cea2e4f4530d4999259ce327f63a89dc043d3d68dR199

I figured it'd be better to be consistent with the approach I take

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 2 above)
Comment (sudeshrshetty) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:206]: it has to be JWT string not 'base64urlencoded'

**Review by Moopli — APPROVED** [review]

File: cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js (hunk 4 above)
Comment (sudeshrshetty) [reply cmd/wallet-js-sdk/src/oidc/presentation/openid4vp.js:246]: vpToken["@context"] = ctx

**Review by sudeshrshetty — APPROVED** [review]

**Pull request merged**